### PR TITLE
FINERACT-1968  - [API] Advanced payment allocation strategy configuration on Loan product

### PR DESCRIPTION
**Pull request by reluxa**
[API] - Advanced payment allocation strategy configuration on Loan product

## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/config/FineractProperties.java [hunk 1]
```diff
@@ -461,6 +461,7 @@ public static class FineractTransactionProcessorProperties {
         private FineractTransactionProcessorItemProperties rbiIndia;
         private FineractTransactionProcessorItemProperties duePenaltyFeeInterestPrincipalInAdvancePrincipalPenaltyFeeInterest;
         private FineractTransactionProcessorItemProperties duePenaltyInterestPrincipalFeeInAdvancePenaltyInterestPrincipalFee;
+        private FineractTransactionProcessorItemProperties advancedRepaymentStrategy;
```
Comment (adamsaghy) [inline fineract-core/src/main/java/org/apache/fineract/infrastructure/core/config/FineractProperties.java:464]: Maybe `payment` would be better.

Comment (adamsaghy) [inline fineract-core/src/main/java/org/apache/fineract/infrastructure/core/config/FineractProperties.java]: Maybe `advancedPaymentStrategy` would be better

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/AdvancedPaymentAllocationsValidator.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.stream.IntStream;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.fineract.infrastructure.core.data.ApiParameterError;
+import org.apache.fineract.infrastructure.core.exception.PlatformApiDataValidationException;
+import org.springframework.stereotype.Service;
+
+@Service
+public class AdvancedPaymentAllocationsValidator {
+
+    public static final String ADVANCED_REPAYMENT_ALLOCATION_STRATEGY = "advanced-repayment-allocation-strategy";
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/AdvancedPaymentAllocationsValidator.java]: Maybe `Repayment` -> `Payment` would be better.

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProduct.java [hunk 3]
```diff
@@ -711,6 +722,12 @@ public LoanProduct(final Fund fund, final String transactionProcessingStrategyCo
     }
 
     public void validateLoanProductPreSave() {
+        if (this.loanProductPaymentAllocationRules != null && loanProductPaymentAllocationRules.size() > 0
+                && !transactionProcessingStrategyCode.equals("advanced-repayment-allocation-strategy")) {
```
Comment (adamsaghy) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProduct.java]: Maybe "advanced-payment-allocation-strategy" would be better.

File: fineract-core/src/main/java/org/apache/fineract/infrastructure/core/config/FineractProperties.java (hunk 1 above)
Comment (reluxa) [reply fineract-core/src/main/java/org/apache/fineract/infrastructure/core/config/FineractProperties.java]: fixed

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/AdvancedPaymentAllocationsValidator.java (hunk 2 above)
Comment (reluxa) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/AdvancedPaymentAllocationsValidator.java]: fixed

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProduct.java (hunk 3 above)
Comment (reluxa) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanproduct/domain/LoanProduct.java]: fixed

**Pull request merged**